Provide extensions for instantiating exceptions

wxbty opened the pull request:
When I use a workflow service as a product and provide services for other applications through rpc, the custom javadelegate class of the exclusive gateway in the workflow xml uploaded by the application cannot be obtained, and the flowable lacks the corresponding extension point to push the instantiation exception to the call square

filiphr commented:
Thanks for the PR @wxbty. Can you provide an example of how this would look like with rpc?

Regarding the extension point, I think it would make more sense to do something like the `DelegateInterceptor`. Basically do an `InstantiateInterceptor`  that looks like:

```
public interface InstantiateInterceptor {

    Object instantiate(String className);

}
```

and then this can be used to instantiate the class.

wxbty commented:
Thanks for your reply, I describe my needs.

We have many business teams that need the help of a flowable workflow engine, but we don't want every application to introduce a flowable jar package and create more than 20 tables. Our current approach is to provide a flowable service (A), and the business application (B) calls the remote interface (dubbo) of A instead of the flowable API, 
and they run like this:
    1. B uploads xml to A and returns processId (0001); 
                 String  pid = remoteUploadService.upload ("xxx.xml");
    2.A deploy (0001);
                 remoteDeployService.deploy (0001);
    3.start process 
     ...
The above is no different from using the flowable api directly, only the difference between local and remote.
But when encountering an exclusive gateway, the problem came. A stored B's xml, but A did not have the JavaDelegate class in the xml, so the process could not go on. Here I hope there are some extension points. When the instantiation of JavaDelegate in xml fails, the class name and context can be sent to B through remote invocation (such as dubbo's callback), so that B can instantiate the class through reflection and continue Business execution continues as defined by the process.
Regarding the naming of the interceptor, I think about it this way, the correct process will go to the interceptor, and exceptions require other things (like spring's filter and exceptionHandler)

dbmalkovsky commented on modules/flowable-engine/src/main/java/org/flowable/engine/impl/cfg/ProcessEngineConfigurationImpl.java:
```diff
@@ -76,12 +76,7 @@
-import org.flowable.common.engine.impl.interceptor.CommandInterceptor;
-import org.flowable.common.engine.impl.interceptor.EngineConfigurationConstants;
-import org.flowable.common.engine.impl.interceptor.SessionFactory;
+import org.flowable.common.engine.impl.interceptor.*;
```
Project standards do not allow package imports; please leave the individual imports.

dbmalkovsky commented on modules/flowable-engine-common/src/main/java/org/flowable/common/engine/impl/interceptor/InstantiateErrorHandler.java:
```diff
@@ -0,0 +1,5 @@
+package org.flowable.common.engine.impl.interceptor;
```
Missing required license header at the top of the file; thanks

filiphr commented:
Thanks for the clarification @wxbty.

Can you please address the other remark I made regarding the `InstantiateInterceptor`. I don't think we need an `InstantiateErrorHandler` if we have the instantiate interceptor.

Regarding your use case, how do you plan to solve transactionality cross machines?

wxbty commented:
Maybe Interceptor is more in line with the overall style of flowable and has been modified. In addition, all write operations are on the flowable service side (A in the use case). I don't think there is a cross-machine transaction problem, thanks

wxbty commented:
> Thanks again for the pr @wxbty.
> 
> I see that you already took some of my input and changed the `InstantiateInterceptor`. Can we please rename the `handle` method into `instantiate` since that's what it does.
> 
> On top of that the current tests should pass without any changes. So I would try to pass the interceptor to the `instantiateDelegate` method instead of only the single place.

Thanks for your suggestion  @filiphr 
Modified as recommended to achieve the purpose of renaming and not modifying test cases. But somehow some test cases can’t pass, it seems to have nothing to do with my code, do you know why?

tijsrademakers commented:
Closing PR because of lack of unit tests. Please reopen when you have time to add unit tests.
